### PR TITLE
Python logger enhancement to support setting log level at real time

### PR DESCRIPTION
**Pull request by Junchao-Mellanox**
# Overview

SONiC provides a Python logger implementation in `sonic_py_common.logger.Logger`. This implementation is a simple wrapper of Python standard package `syslog`. It does not provide the ability to change log level at real time. Sometimes, in order to get more debug information, developer has to manually change the log level in code on a running switch and restart the Python daemon. This PR introduces an enhancement to allow user setting log level on fly.

# PR

TBD

### CODE REVIEW

**Review by keboliu — APPROVED** [review]

**Comment by liat-grozovik** [comment]
@prgeor appreciate your feedback on the HLD

File: doc/syslog/python-logger-enhancement.md
Comment (prgeor) [inline doc/syslog/python-logger-enhancement.md:64]: @Junchao-Mellanox backward compatible with ?

File: doc/syslog/python-logger-enhancement.md
Comment (prgeor) [inline doc/syslog/python-logger-enhancement.md:1]: @Junchao-Mellanox won't there will be conflict with set_min_log_priority_info() set by the particular daemon and user overrides with real time debug level and forgets to set the log level to default as per daemon init?

File: doc/syslog/python-logger-enhancement.md
Comment (Junchao-Mellanox) [reply doc/syslog/python-logger-enhancement.md:64]: backward compatible with its original interface. I will re-phrase.

Comment (Junchao-Mellanox) [reply doc/syslog/python-logger-enhancement.md]: Fixed.

File: doc/syslog/python-logger-enhancement.md
Comment (Junchao-Mellanox) [reply doc/syslog/python-logger-enhancement.md:1]: Hi @prgeor , you are right. There could be conflict if user forgets to set the log level to default value. One possible solution is to make "set_min_log_priority_info" directly write CONFIG DB if enable_set_log_level_on_fly is True. Do you want me to change it like that? Or, do you have other suggestions?

Comment (prgeor) [reply doc/syslog/python-logger-enhancement.md:1]: @Junchao-Mellanox how does writing to CONFIG_DB prevent the issue I said?

Comment (Junchao-Mellanox) [reply doc/syslog/python-logger-enhancement.md:1]: Hi @prgeor , I think the conflict you mentioned is like this:
```
1. User create logger instance
2. User code set logger level to INFO (But DB has logger level DEBUG)
```

In step#2, there could be a race condition that: code wants to set logger level to INFO in main thread, DB thread wants to set logger level to DEBUG level. It depends on which code runs last. I think this is the conflict.

And if we implement   set_min_log_priority_info() like this:
```
def set_min_log_priority_info(level):
    if self.enable_set_log_level_on_fly:
        # do nothing, let DB thread handle the log level
    else:
       # directly set log level
```

In this case, we will have consistent result. What do you think?

Comment (prgeor) [reply doc/syslog/python-logger-enhancement.md:1]: @Junchao-Mellanox so we need to protect `enable_set_log_level_on_fly` from race condition too?

Comment (Junchao-Mellanox) [reply doc/syslog/python-logger-enhancement.md:1]: that flag is set only via __init__, no need to protect it.

**Comment by liat-grozovik** [comment]
@prgeor could you please refer to the latest doc following feedback and see if additional comments available?
@keboliu can you confirm all relevant PRs are tagged on this PR description?

**Comment by prgeor** [comment]
> @prgeor could you please refer to the latest doc following feedback and see if additional comments available? @keboliu can you confirm all relevant PRs are tagged on this PR description?

@liat-grozovik ack

**Comment by prgeor** [comment]
@Junchao-Mellanox There is one flaw with existing implementation of Logging in `sonic_py_common/logger.py` where syslog() call impacts all the daemons running in the container. To understand what I am saying, take a look at this PR which is trying to fix this https://github.com/sonic-net/sonic-buildimage/pull/17171/files

We need log level which is local to that daemon which is logging. The existing Logger implementation pollutes the log identifier of other daemons as well.

**Comment by Junchao-Mellanox** [comment]
Hi @liat-grozovik,

This PR is pending on PR https://github.com/sonic-net/sonic-buildimage/pull/17171 which is handled by @xincunli-sonic . For PR 17171, we have one open item:

- log message always contain "localhost" instead of real hostname. 

For this item, I can always reproduce on my side, but Xincun cannot reproduce on his side. So, I requested a debug session with him this week. I am waiting for his response.

**Comment by liat-grozovik** [comment]
thanks @Junchao-Mellanox for keeping handling it. now that the pending PR is merged, i assume you need new PR on top of it. once you have please update the PR description with the new code PR instead of the closed one.

@prgeor can we get your approval for the HLD?

**Comment by prgeor** [comment]
> thanks @Junchao-Mellanox for keeping handling it. now that the pending PR is merged, i assume you need new PR on top of it. once you have please update the PR description with the new code PR instead of the closed one.
> 
> @prgeor can we get your approval for the HLD?

@liat-grozovik reviewing

File: doc/syslog/python-logger-enhancement.md
Comment (prgeor) [inline doc/syslog/python-logger-enhancement.md:22]: @Junchao-Mellanox How are you overcoming this limitation? I see that you are still writing to the DB which means redis server needs to be running

**Review by prgeor — COMMENTED** [review]
@Junchao-Mellanox to keep things simple, we an make use of config_db logger table for user to configure the log level for debugging purpose like below:-

```
 "LOGGER": {
            "xcvrd": {
                "LOGLEVEL": "DEBUG",
                "LOGOUTPUT": "SYSLOG"
            },
```

Use the Identifier specified in syslogger to fetch the current log level from config DB 
For eg. The identfier could be Xcvrd, CmisMgr, StateUpdateTask etc.

The advantage is we don't need to create extra threads for handling this

Reading redis DB is not a costly affair

This would limit the change to syslogger 

Please note, logger.py will be deprecated over time as its usage causes confusion and because of which we introduced syslogger

**Comment by prgeor** [comment]
@Junchao-Mellanox one main advantage of the above approach is enabling logging at module level. For eg. Xcvrd common or main thread can have one SysLogger("Xcvrd") instance for eg. CMIS mgr may have another SysLogger("CmisMgr") instance. The logging can be enabled per module within the same thread/process

File: doc/syslog/python-logger-enhancement.md
Comment (Junchao-Mellanox) [reply doc/syslog/python-logger-enhancement.md:22]: To overcome it, the implementation will:
1. At the beginning of the thread function, it will wait until redis is up. If the connect to redis fails, it will silently ignore the connection error and retry in 5 seconds. 
2. In the while loop of thread function, catch connection error while reading from redis

**Comment by Junchao-Mellanox** [comment]
> @Junchao-Mellanox to keep things simple, we an make use of config_db logger table for user to configure the log level for debugging purpose like below:-
> 
> ```
>  "LOGGER": {
>             "xcvrd": {
>                 "LOGLEVEL": "DEBUG",
>                 "LOGOUTPUT": "SYSLOG"
>             },
> ```
> 
> Use the Identifier specified in syslogger to fetch the current log level from config DB For eg. The identfier could be Xcvrd, CmisMgr, StateUpdateTask etc.
> 
> The advantage is we don't need to create extra threads for handling this
> 
> Reading redis DB is not a costly affair
> 
> This would limit the change to syslogger
> 
> Please note, logger.py will be deprecated over time as its usage causes confusion and because of which we introduced syslogger

Hi @prgeor , this is exactly the implementation will do.

**Comment by Junchao-Mellanox** [comment]
> @Junchao-Mellanox one main advantage of the above approach is enabling logging at module level. For eg. Xcvrd common or main thread can have one SysLogger("Xcvrd") instance for eg. CMIS mgr may have another SysLogger("CmisMgr") instance. The logging can be enabled per module within the same thread/process

Totally agree. And this is exactly my implementation plan.

**Comment by prgeor** [comment]
> > @Junchao-Mellanox one main advantage of the above approach is enabling logging at module level. For eg. Xcvrd common or main thread can have one SysLogger("Xcvrd") instance for eg. CMIS mgr may have another SysLogger("CmisMgr") instance. The logging can be enabled per module within the same thread/process
> 
> Totally agree. And this is exactly my implementation plan.

@Junchao-Mellanox why do we need a thread?

File: doc/syslog/python-logger-enhancement.md
Comment (prgeor) [inline doc/syslog/python-logger-enhancement.md:155]: @Junchao-Mellanox if user does not specify the optional argument --service, then also this will work?

Comment (Junchao-Mellanox) [reply doc/syslog/python-logger-enhancement.md:155]: If `--service` is not present, the command will try to send the signal to host side process which has the pid equals to `--pid`

File: doc/syslog/python-logger-enhancement.md
Comment (prgeor) [inline doc/syslog/python-logger-enhancement.md:182]: @Junchao-Mellanox can we have option `signal` instead of `level` to send signal?

Comment (Junchao-Mellanox) [reply doc/syslog/python-logger-enhancement.md:182]: Hi @prgeor , the requirement is to have one single command which update log level and make it effect, so, I have to add these two function together.

File: doc/syslog/python-logger-enhancement.md
Comment (prgeor) [inline doc/syslog/python-logger-enhancement.md:105]: @Junchao-Mellanox Why we need per process handling? The signal handling is per process so child process is independ of parent process and can have its own signal handling. Probably what we need is to support single process with multiple threads.

Comment (Junchao-Mellanox) [reply doc/syslog/python-logger-enhancement.md:105]: Hi @prgeor , this is just an example in case user uses it in multi-processing case. Your suggestion is also a good option and probably a more general option for multi processing case. Basically, this feature supports single process with multiple threads well.

Comment (Junchao-Mellanox) [reply doc/syslog/python-logger-enhancement.md]: fixed

File: doc/syslog/python-logger-enhancement.md
Comment (prgeor) [inline doc/syslog/python-logger-enhancement.md:55]: @Junchao-Mellanox can we maintain the SysLogger Objects inside src/sonic-py-common/sonic_py_common/daemon_base.py so that the main thread will create the list of SysLogger objects required in `DaemonBase.logger_instances[]`? This way the main daemon process will be able to do the signal handling based upon the `identifier` assigned to each Syslogger debug level configured in the CONFIG_DB

**Comment by liat-grozovik** [comment]
@prgeor could you please help to review following the recent changes based on your feedback?

**Comment by liat-grozovik** [comment]
@Junchao-Mellanox is there code PR aligned now with the updated HLD? if so, can you please remove the closed PR and update the description with the relevant ones?

**Comment by Junchao-Mellanox** [comment]
Hi @liat-grozovik , I will remove the closed one. I will add the code PR once it is ready for review.

File: doc/syslog/python-logger-enhancement.md
Comment (prgeor) [inline doc/syslog/python-logger-enhancement.md:42]: @Junchao-Mellanox Why syslogger need to save log level to DB? The user will configure the log level which will be updated to DB and the main process would be woken up via signal to fetch the user set log level from the DB. There is no need to save log level within SysLogger class.

File: doc/syslog/python-logger-enhancement.md
Comment (prgeor) [inline doc/syslog/python-logger-enhancement.md:53]: @Junchao-Mellanox why `enable_runtime_config` default is kept False? What is the intention?

File: doc/syslog/python-logger-enhancement.md
Comment (prgeor) [inline doc/syslog/python-logger-enhancement.md:55]: @Junchao-Mellanox can we rename the method to `update_log_level` ?

File: doc/syslog/python-logger-enhancement.md
Comment (Junchao-Mellanox) [reply doc/syslog/python-logger-enhancement.md:42]: we need to register logger itself to DB so that `swssloglevel -p` will print all the available log identifiers. Like this:
```
swssloglevel -p
SAI_API_ACL                   SAI_LOG_LEVEL_NOTICE
SAI_API_ARS                   SAI_LOG_LEVEL_NOTICE
SAI_API_ARS_PROFILE           SAI_LOG_LEVEL_NOTICE
SAI_API_BFD                   SAI_LOG_LEVEL_NOTICE
SAI_API_BMTOR                 SAI_LOG_LEVEL_NOTICE
SAI_API_BRIDGE                SAI_LOG_LEVEL_NOTICE
SAI_API_BUFFER                SAI_LOG_LEVEL_NOTICE
SAI_API_COUNTER               SAI_LOG_LEVEL_NOTICE
SAI_API_DASH_ACL              SAI_LOG_LEVEL_NOTICE
SAI_API_DASH_DIRECTION_LOOKUP SAI_LOG_LEVEL_NOTICE
SAI_API_DASH_ENI              SAI_LOG_LEVEL_NOTICE
SAI_API_DASH_INBOUND_ROUTING  SAI_LOG_LEVEL_NOTICE
SAI_API_DASH_METER            SAI_LOG_LEVEL_NOTICE
SAI_API_DASH_OUTBOUND_CA_TO_PASAI_LOG_LEVEL_NOTICE
SAI_API_DASH_OUTBOUND_ROUTING SAI_LOG_LEVEL_NOTICE
SAI_API_DASH_PA_VALIDATION    SAI_LOG_LEVEL_NOTICE
SAI_API_DASH_VIP              SAI_LOG_LEVEL_NOTICE
SAI_API_DASH_VNET             SAI_LOG_LEVEL_NOTICE
SAI_API_DEBUG_COUNTER         SAI_LOG_LEVEL_NOTICE
SAI_API_DTEL                  SAI_LOG_LEVEL_NOTICE
SAI_API_FDB                   SAI_LOG_LEVEL_NOTICE
SAI_API_GENERIC_PROGRAMMABLE  SAI_LOG_LEVEL_NOTICE
SAI_API_HASH                  SAI_LOG_LEVEL_NOTICE
SAI_API_HOSTIF                SAI_LOG_LEVEL_NOTICE
SAI_API_IPMC                  SAI_LOG_LEVEL_NOTICE
SAI_API_IPMC_GROUP            SAI_LOG_LEVEL_NOTICE
SAI_API_IPSEC                 SAI_LOG_LEVEL_NOTICE
SAI_API_ISOLATION_GROUP       SAI_LOG_LEVEL_NOTICE
SAI_API_L2MC                  SAI_LOG_LEVEL_NOTICE
SAI_API_L2MC_GROUP            SAI_LOG_LEVEL_NOTICE
SAI_API_LAG                   SAI_LOG_LEVEL_NOTICE
SAI_API_MACSEC                SAI_LOG_LEVEL_NOTICE
SAI_API_MCAST_FDB             SAI_LOG_LEVEL_NOTICE
SAI_API_MIRROR                SAI_LOG_LEVEL_NOTICE
SAI_API_MPLS                  SAI_LOG_LEVEL_NOTICE
SAI_API_MY_MAC                SAI_LOG_LEVEL_NOTICE
SAI_API_NAT                   SAI_LOG_LEVEL_NOTICE
SAI_API_NEIGHBOR              SAI_LOG_LEVEL_NOTICE
SAI_API_NEXT_HOP              SAI_LOG_LEVEL_NOTICE
SAI_API_NEXT_HOP_GROUP        SAI_LOG_LEVEL_NOTICE
SAI_API_POLICER               SAI_LOG_LEVEL_NOTICE
SAI_API_PORT                  SAI_LOG_LEVEL_NOTICE
SAI_API_QOS_MAP               SAI_LOG_LEVEL_NOTICE
SAI_API_QUEUE                 SAI_LOG_LEVEL_NOTICE
SAI_API_ROUTE                 SAI_LOG_LEVEL_NOTICE
SAI_API_ROUTER_INTERFACE      SAI_LOG_LEVEL_NOTICE
SAI_API_RPF_GROUP             SAI_LOG_LEVEL_NOTICE
SAI_API_SAMPLEPACKET          SAI_LOG_LEVEL_NOTICE
SAI_API_SCHEDULER             SAI_LOG_LEVEL_NOTICE
SAI_API_SCHEDULER_GROUP       SAI_LOG_LEVEL_NOTICE
SAI_API_SRV6                  SAI_LOG_LEVEL_NOTICE
SAI_API_STP                   SAI_LOG_LEVEL_NOTICE
SAI_API_SWITCH                SAI_LOG_LEVEL_NOTICE
SAI_API_SYSTEM_PORT           SAI_LOG_LEVEL_NOTICE
SAI_API_TAM                   SAI_LOG_LEVEL_NOTICE
SAI_API_TUNNEL                SAI_LOG_LEVEL_NOTICE
SAI_API_TWAMP                 SAI_LOG_LEVEL_NOTICE
SAI_API_UDF                   SAI_LOG_LEVEL_NOTICE
SAI_API_VIRTUAL_ROUTER        SAI_LOG_LEVEL_NOTICE
SAI_API_VLAN                  SAI_LOG_LEVEL_NOTICE
SAI_API_WRED                  SAI_LOG_LEVEL_NOTICE
buffermgrd                    NOTICE
coppmgrd                      NOTICE
fdbsyncd                      NOTICE
fpmsyncd                      NOTICE
gearsyncd                     NOTICE
intfmgrd                      NOTICE
nbrmgrd                       NOTICE
neighsyncd                    NOTICE
orchagent                     NOTICE
portmgrd                      NOTICE
portsyncd                     NOTICE
sflowmgrd                     NOTICE
syncd                         NOTICE
teammgrd                      NOTICE
teamsyncd                     NOTICE
tlm_teamd                     NOTICE
tunnelmgrd                    NOTICE
vlanmgrd                      NOTICE
vrfmgrd                       NOTICE
vxlanmgrd                     NOTICE
wjhcli                        NOTICE
wjhd                          NOTICE
```

File: doc/syslog/python-logger-enhancement.md
Comment (Junchao-Mellanox) [reply doc/syslog/python-logger-enhancement.md:53]: Two reason here:
1. We want to keep backward compatible. Before the feature, there is no runtime config support for SysLogger.
2. Enabling runtime config will slow down the logger creation for a litter bit, we want each SysLogger instance to only enable it when need.

File: doc/syslog/python-logger-enhancement.md
Comment (Junchao-Mellanox) [reply doc/syslog/python-logger-enhancement.md:55]: sure

**Review by prgeor — APPROVED** [review]

**Pull request merged**
